[RFC] vim-patch.sh: Improvements for review functionality, submission option.

fwalch opened the pull request:
`vim-patch.sh -r`
- Support pull requests with multiple commits.
- Offer to clean up files after review.
- Always use full commit hash in assign_commit_details().

`vim-patch.sh -s`
- When calling "vim-patch.sh -s" on a checked-out branch created with
  "vim-patch.sh -p", push to origin,
  create pull request for all commits on branch (using hub), and clean up patch files.

`vim-patch.sh -p`
- If already on a checked-out branch created from a previous "vim-patch.sh -p", just create commit (don't create additional new branch).

fwalch committed:
vim-patch.sh: Improvements for review functionality.

 * Support pull requests with multiple commits.
 * Offer to clean up files after review.
 * Always use full commit hash in assign_commit_details().

Pyrohh commented on scripts/vim-patch.sh:
```diff
@@ -35,6 +37,27 @@ check_executable() {
+  read -p "Delete these files (Y/n)? " -n 1 -r reply
+  echo
+  if [[ "${reply}" =~ ^[Yy]$ ]]; then
+    rm ${CREATED_FILES}
```
`rm` -> `rm --` ?

fwalch replied:
:+1: can't hurt. Also, there was actually an error here; changed `${CREATED_FILES}` to `${CREATED_FILES[@]}`.

Pyrohh commented on scripts/vim-patch.sh:
```diff
@@ -35,6 +37,27 @@ check_executable() {
+  if [[ "${reply}" =~ ^[Yy]$ ]]; then
+    rm ${CREATED_FILES}
+  else
+    echo "You can use 'git clean' to remove these files when you're done."
```
`git clean` would remove a users `local.mk`, so maybe it could instead be:
```
echo "To remove these files when you're done:"
echo "  rm ${CREATED_FILES}"

fwalch replied:
`local.mk` is in `.gitignore`, so it shouldn't get removed.

justinmk commented:
Can we instead just output only the `.patch` file? The `.diff` and `.pr` files--particularly for the `-p` option of `vim-patch.sh`--are just confusing and mostly redundant. 

- We wouldn't need these ornate script features if we just output .patch files: user can just `rm vim*.patch`.
- The extra files are mostly just confusing and not very useful for people trying to create patches. They must inspect the contents of the file and then wonder "why do I need these files"

fwalch commented:
@justinmk the `.pr` file is for the PR description, so I'd argue that's useful. But sure, just outputting `.patch` files and omitting `.diff` should be fine.

justinmk commented:
I humbly suggest that the PR description doesn't need the original patch in it. A link to the original patch is more than enough--I don't trust the PR description patch anyways.

justinmk commented:
At some point it would be nice to have an option to auto-create the PR for the contributor. That also avoids the extra `.pr` file.

fwalch commented:
> I humbly suggest that the PR description doesn't need the original patch in it. A link to the original patch is more than enough--I don't trust the PR description patch anyways.

Fair enough.. I think that might be a remnant from the Google Code times, where (AFAIR) no function names were in the diff and generally GitHub's way of formatting them was much more readable.

> At some point it would be nice to have an option to auto-create the PR for the contributor. That also avoids the extra .pr file.

I thought about that previously, but held off because that can be done with `hub pull-request -F xxx.pr`.

fwalch committed:
vim-patch.sh: Use .patch files only.

fwalch committed:
vim-patch.sh: Add -s (submit pull request) option.

When calling "vim-patch.sh -s" on a checked-out branch created with
"vim-patch.sh -p", create commit from staged changes, push to origin,
create pull request (using hub), and clean up patch files.

fwalch commented:
Added a `-s` option to submit a pull request.

fwalch commented:
@watiko @jbradaric @jusga What do you think of these changes to `vim-patch.sh`? See https://github.com/neovim/neovim/pull/4079#issuecomment-174220376 for how it would work. Would this make it easier for you to port Vim patches? Any other suggestions on how to improve the script?

watiko commented:
@fwalch Hmm, I want the option which only submits pull request. BTW the script does not contain any linting, though it is not always needed.

fwalch commented:
> Hmm, I want the option which only submits pull request.

Okay, so `-p` should continue creating the empty commit with the correct message.

I don't want to complicate all of this too much by adding more options.. But, to help with multi-patch PRs, the easiest way would is to detect if there's already a `vim-...` branch checked out. Example:

* ./scripts/vim-patch.sh -p 7.4.123
  * fetch upstream/master
  * create & check out branch vim-7.4.123 based on upstream/master
  * create empty commit with correct message
  * user ports patch
* ./scripts/vim-patch.sh -p 7.4.124
  * creates empty commit with correct message
  * user ports patch
* ./scripts/vim-patch.sh -s
  * create pull request message from git diff upstream/master..HEAD
  * submit pull request to GitHub

> BTW the script does not contain any linting, though it is not always needed.

Linting is a good point.. although I am not sure if it should be added to the script to run automatically, it takes a really long time (over 40 seconds).

fwalch committed:
vim-patch.sh: Support multi-patch pull requests.

* Calling "vim-patch.sh -p" on a checked-out branch already created with
  "-p" will re-use the branch and append commits.

* Fetch upstream/master before checking out branch on first call of "-p".

* Reverted creation of commit in submit step ("-s") to previous behavior:
  Create an empty commit with correct commit message when "-p" is called.

* Submitting a pull request with "-s" will create a correct pull request
  message even if multiple patches have been ported in one single branch
  with "-p".

watiko commented:
@fwalch  LGTM.

> it takes a really long time (over 40 seconds).

Yes...

jbradaric commented:
@fwalch LGTM. This would make it much easier for me to create multi-patch PRs.

fwalch commented:
@justinmk Do you have more suggestions? Otherwise I think I'd merge this :-)

watiko commented:
> Should not we do something with “closes #470”? Our #470 has nothing to do with what is referenced in the commit message.
> 
> I think that vim-patch.sh should at least change this to vim/vim#470.

As @ZyX-I  said (https://github.com/neovim/neovim/pull/4265#issuecomment-184253955), we should treat `#number` correctly.

fwalch commented:
@watiko Updated.

fwalch committed:
vim-patch.sh: Replace #<number> with vim/vim#<number> in commit messages.

watiko commented:
:+1:

justinmk commented:
Thanks for this @fwalch , the changes are nice! The `-r` feature seems to have some quirks for me, but in the meantime this is an improvement so let's not hold it up.
